group-coordinator/src/main/java/org/apache/kafka/coordinator/group/generic/GenericGroup.java

```diff
@@ -849,6 +853,42 @@ public void validateOffsetFetch(
         }
     }
 
+    /**
+     * Validates the OffsetDelete request.
+     */
+    @Override
+    public void validateOffsetDelete() throws ApiException {
+        if (isInState(DEAD)) {
```

Comment:
@dajac the current [draft PR for group/offset expiration](https://github.com/apache/kafka/pull/14467) transitions groups to dead once a group is empty && offsets are gone.

the current behavior for generic groups is the above, and i copied the same behavior for consumer groups. then once the group is Dead, it will be considered for expiration in the next cycle.

Reply:
I am not sure to understand why we need to do this. Couldn't we just delete the group when it is empty and offsets are gone instead of transitioning to Dead and then deleting it?

My understanding is that we use Dead in the old code because we can't remove the group from the map before the change is committed to the log. During this time, the group is in the Dead state. In our world, the group is remove from the map immediately and the change is reverted if the write fails.